### ISSUE TITLE
pod-checkpointer ignoring registry-mirror

### ISSUE THREAD

**Issue by jsimonetti** [talos-systems/talos#1740]
While trying to get Talos airgapped, the last hurdle is pod-checkpointer. Somehow, the registry-mirror configuration is ignore and it is always pulled from internet.

Repro steps:
- Create a local registry with:
```
version: '2'

services:
  registry:
    container_name: docker-registry
    restart: always
    image: registry:2
    environment:
      REGISTRY_HTTP_TLS_CERTIFICATE: /certs/domain.crt
      REGISTRY_HTTP_TLS_KEY: /certs/domain.key
      REGISTRY_STORAGE_DELETE_ENABLED: "true"
      REGISTRY_LOG_ACCESSLOG_DISABLED: "false"
    volumes:
      - ./data:/var/lib/registry
      - ./certs:/certs
    networks:
      registry:
        ipv4_address: 10.11.0.2

networks:
  registry:
    driver: bridge
    driver_opts:
      com.docker.network.bridge.enable_icc: "true"
      com.docker.network.bridge.name: "registry0"
    ipam:
      config:
        - subnet: 10.11.0.0/24
          gateway: 10.11.0.1
```
- Add an iptables rule to allow 10.5.0.0/24 to access 10.11.0.2:5000 if applicable, (exercise up to reader)
- Add the registry to local docker config:
```
  "insecure-registries": ["10.11.0.2:5000"],
  "registry-mirrors": ["https://10.11.0.2:5000"]
```
- Load the registry with the correct containers:
```
docker pull k8s.gcr.io/hyperkube:v1.17.0
docker tag k8s.gcr.io/hyperkube:v1.17.0 10.11.0.2:5000/hyperkube:v1.17.0
docker push 10.11.0.2:5000/hyperkube:v1.17.0
docker pull k8s.gcr.io/etcd:3.3.15-0
docker tag k8s.gcr.io/etcd:3.3.15-0 10.11.0.2:5000/etcd:3.3.15-0
docker push 10.11.0.2:5000/etcd:3.3.15-0
docker pull k8s.gcr.io/pause:3.1
docker tag k8s.gcr.io/pause:3.1 10.11.0.2:5000/pause:3.1
docker push 10.11.0.2:5000/pause:3.1
docker pull k8s.gcr.io/coredns:1.6.5
docker tag k8s.gcr.io/coredns:1.6.5 10.11.0.2:5000/coredns:1.6.5
docker push 10.11.0.2:5000/coredns:1.6.5
docker pull quay.io/coreos/flannel:v0.11.0-amd64
docker tag quay.io/coreos/flannel:v0.11.0-amd64 10.11.0.2:5000/coreos/flannel:v0.11.0-amd64
docker push 10.11.0.2:5000/coreos/flannel:v0.11.0-amd64
docker pull quay.io/coreos/flannel-cni:v0.3.0
docker tag quay.io/coreos/flannel-cni:v0.3.0 10.11.0.2:5000/coreos/flannel-cni:v0.3.0
docker push 10.11.0.2:5000/coreos/flannel-cni:v0.3.0
docker pull docker.io/autonomy/pod-checkpointer:latest
docker tag docker.io/autonomy/pod-checkpointer:latest 10.11.0.2:5000/autonomy/pod-checkpointer:latest
docker push 10.11.0.2:5000/autonomy/pod-checkpointer:latest
docker pull docker.io/autonomy/installer:latest
docker tag docker.io/autonomy/installer:latest 10.11.0.2:5000/autonomy/installer:latest
docker push 10.11.0.2:5000/autonomy/installer:latest

```
- Create a new cluster with `osctl generate`
- Modify the .yaml files using:
```
    yq w -i -- init.yaml 'machine.files[+].content' "$(cat ~/dev/registry/certs/domain.crt)"
    yq w -i init.yaml 'machine.files[0].permissions' '0644'
    yq w -i init.yaml 'machine.files[0].path' '/etc/ssl/certs/ca-certificates'
    yq w -i init.yaml 'machine.files[0].op' 'append'
    yq w -i -- init.yaml 'machine.files[+].content' "$(cat container-registry.toml)"
    yq w -i init.yaml 'machine.files[1].permissions' '0644'
    yq w -i init.yaml 'machine.files[1].path' '/etc/cri/containerd.toml'
    yq w -i init.yaml 'machine.files[1].op' 'append'
    #FOR KVM: yq w -i init.yaml machine.install.image '10.11.0.2:5000/autonomy/installer:latest'
    yq w -i init.yaml cluster.etcd.image '10.11.0.2:5000/etcd:3.3.15-0'
    yq w -i init.yaml machine.kubelet.image '10.11.0.2:5000/hyperkube:v1.17.0'

    yq w -i -- controlplane.yaml 'machine.files[+].content' "$(cat ~/dev/registry/certs/domain.crt)"
    yq w -i controlplane.yaml 'machine.files[0].permissions' '0644'
    yq w -i controlplane.yaml 'machine.files[0].path' '/etc/ssl/certs/ca-certificates'
    yq w -i controlplane.yaml 'machine.files[0].op' 'append'
    yq w -i -- controlplane.yaml 'machine.files[+].content' "$(cat container-registry.toml)"
    yq w -i controlplane.yaml 'machine.files[1].permissions' '0644'
    yq w -i controlplane.yaml 'machine.files[1].path' '/etc/cri/containerd.toml'
    yq w -i controlplane.yaml 'machine.files[1].op' 'append'
    #FOR KVM: yq w -i controlplane.yaml machine.install.image '10.11.0.2:5000/autonomy/installer:latest'
    yq w -i controlplane.yaml cluster.etcd.image '10.11.0.2:5000/etcd:3.3.15-0'
    yq w -i controlplane.yaml machine.kubelet.image '10.11.0.2:5000/hyperkube:v1.17.0'

    yq w -i -- join.yaml 'machine.files[+].content' "$(cat ~/dev/registry/certs/domain.crt)"
    yq w -i join.yaml 'machine.files[0].permissions' '0644'
    yq w -i join.yaml 'machine.files[0].path' '/etc/ssl/certs/ca-certificates'
    yq w -i join.yaml 'machine.files[0].op' 'append'
    yq w -i -- join.yaml 'machine.files[+].content' "$(cat container-registry.toml)"
    yq w -i join.yaml 'machine.files[1].permissions' '0644'
    yq w -i join.yaml 'machine.files[1].path' '/etc/cri/containerd.toml'
    yq w -i join.yaml 'machine.files[1].op' 'append'
    #FOR KVM: yq w -i join.yaml machine.install.image '10.11.0.2:5000/autonomy/installer:latest'
    yq w -i join.yaml machine.kubelet.image '10.11.0.2:5000/hyperkube:v1.17.0'
```

- Where:
    - 10.11.0.2:5000 is the local registry mirror address
    - ~/dev/registry/certs/domain.crt is the certificate for the local registry mirror
    - `container-registry.toml` is a file containing:
```
[plugins.cri.registry.mirrors]
  [plugins.cri.registry.mirrors."*"]
    endpoint = ["https://10.11.0.2:5000"]
[plugins.cri.registry.configs."10.11.0.2:5000".tls]
  insecure_skip_verify = true
```
- Airgap your installation (unplug cable or do something like `iptables -I FORWARD -s 10.5.0.0/24 -j REJECT`

Watching logs on console or with `docker logs -f master-1` or with `osctl logs -f containerd` should show you an erro when/if pod-checkpointer is downloaded. (depending on airgap method, this will be an error resolving `docker.io` dns record for the above method)

**Comment by jsimonetti** [comment]
On a side note, I wish it would be enough to just add the `/etc/cri/containerd.toml` override. Unfortunately, Talos (or bootkube, I am unsure which) ignores that file and always pull from internet. This makes it necessary to manually change all images used, which is quite the churn.

**Comment by jsimonetti** [comment]
#1744 will probably fill this last gap and allow me to override the pod-checkpointer image to `10.11.0.2:5000/autonomy/pod-checkpointer:latest`

**Comment by rsmitty** [comment]
Closed via talos-systems/talos#1799 

@jsimonetti reopen if that's not the case!

**Comment by jsimonetti** [comment]
I agree #1799 is a workaround. However, that does not really fix the issue mentioned here. This issue might pop up later and forces someone to change the image urls **aswell** as `/etc/cri/containerd.toml`.
I would agree that this is not a high priority, since there is a workaround available, but still would consider this a bug; especially since the other bootkube images are pulled correctly and only pod-checkpointer appears to be affected.
